### DISCUSSION TITLE
Why there is no gcc in the image despite the fact that the installation command is visible in the layers?

### DISCUSSION THREAD

**Discussion by ant0nk** [apache/airflow#27440] [Q&A]
There is a command in one of the layers which among other things installs gcc - 
`/bin/sh -c set -eux; 		savedAptMark="$(apt-mark showmanual)"; 	apt-get update; 	apt-get install -y --no-install-recommends 		dpkg-dev 		gcc...`
but there is no gcc inside image, so when I do `docker run -it apache/airflow:2.2.4-python3.8 bash -c "gcc"` I see only `/bin/bash: gcc: command not found`
I need gcc because I'm trying to make custom image with providers requiring gcc to be installed.

**Comment by Taragolis** [comment]
This command from python slim-buster image
https://github.com/docker-library/python/blob/230bb8c07e599a1e230b63895647e677d109c5c0/3.8/slim-buster/Dockerfile#L29-L54

I do not investigate entire command but most probably `gcc` removed in the final `python:3.8-slim-buster` image which uses as base image for Airflow (python 3.8).

So you have at least two options:
1. [Extend airflow image](https://airflow.apache.org/docs/docker-stack/build.html#building-the-image)
2. [Build custom airflow image](https://airflow.apache.org/docs/docker-stack/build.html#customizing-the-image)

**Comment by potiuk** [comment]
Airflow Docker file follows best practices of image building.

It is a multi-segmented image (following the best practices of making images "slim"). There is a build segment - where gcc (and build-essentials) are present and the final segment only contains already runtime components (saves about 200 MB of size of the image).

This chapter explains it all in detail - https://airflow.apache.org/docs/docker-stack/build.html#extending-vs-customizing-the-image including examples showing how much is saved. Just read the dosc  @ant0nk. You even have some examples of how you can add a package which require "build-essentials/gcc" - it shows several cases and discusses advantages and disadvantages of those approaches.

Happy reading :).
